### PR TITLE
[IMAGING-316] Add support for BigTIFF format

### PR DESCRIPTION
**Pull request by gwlucastrig**
These changes add support for the BigTIFF format. They enable the Commons Imaging library to read TIFF files of size larger than supported by a 32-bit address space.  A dedicated test case and a set of BigTIFF sample files are included.

Please see the Commons Imaging JIRA tracker issue 316 at  [https://issues.apache.org/jira/browse/IMAGING-316](https://issues.apache.org/jira/browse/IMAGING-316) for more detail.

The ability to write BigTIFF files will be reserved for a future pull request

### CODE REVIEW

**Comment by gwlucastrig** [comment]
Do you have a feel for when this PR might get merged or, failing that, rejected?  I've got some other stuff I'm working on and am trying to plan for the next few weeks.

File: src/main/java/org/apache/commons/imaging/common/BinaryFunctions.java [hunk 1]
```diff
@@ -177,6 +177,44 @@ public static int read4Bytes(final String name, final InputStream is,
         return result;
     }
 
+   public static long read8Bytes(final String name,
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/common/BinaryFunctions.java:180]: Javadoc?

**Review by garydgregory — CHANGES_REQUESTED** [review]
See my 2 comments. In general, I would be really nice if new public and protected elements carried Javadocs. That will help us going forward I think.

**Comment by gwlucastrig** [comment]
Will address these comments and update code.

**Comment by gwlucastrig** [comment]
I hope I did this right.  Please let me know if it didn't work. 

I am not adept at Git and was unable to find a clear set of instructions for submitting changes in response to a PR review. I'm sure they're out there, but I didn't find them.

File: src/main/java/org/apache/commons/imaging/common/ByteConversions.java [hunk 2]
```diff
@@ -399,6 +399,43 @@ private static int[] toUInt16s(final byte[] bytes, final int offset, final int l
         return result;
     }
 
+     public static long toLong(final byte[] bytes, final ByteOrder byteOrder) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/common/ByteConversions.java:402]: We should provide Javadocs on all new public and protected methods IMO.

Comment (garydgregory) [reply src/main/java/org/apache/commons/imaging/common/ByteConversions.java:402]: Also, it looks like there is an extra space in front of `public`?

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java [hunk 3]
```diff
@@ -234,6 +234,14 @@ public int[] getIntArrayValue() throws ImagingException {
             final int[] numbers = (int[]) o;
             return Arrays.copyOf(numbers, numbers.length);
         }
+        if (o instanceof long[]){
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:237]: The formatting seems off in the block, missing spaces.

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java [hunk 4]
```diff
@@ -288,6 +296,55 @@ public int getIntValueOrArraySum() throws ImagingException {
         // return -1;
     }
 
+    public long[] getLongArrayValue() throws ImagingException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:299]: We should provide Javadocs on all new public and protected methods IMO.

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java [hunk 5]
```diff
@@ -288,6 +296,55 @@ public int getIntValueOrArraySum() throws ImagingException {
...
+          final long[] numbers = (long[]) o;
+          return Arrays.copyOf(numbers, numbers.length);
+        }
+
+        throw new ImagingException("Unknown value: " + o + " for: "
+                + getTagInfo().getDescription());
+        // return null;
+    }
+
+    public long getLongValue() throws ImagingException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:335]: We should provide Javadocs on all new public and protected methods IMO.

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java [hunk 6]
```diff
@@ -455,27 +473,47 @@ private TiffHeader readTiffHeader(final ByteSource byteSource) throws ImagingExc
...
-        }
-
-        final ByteOrder byteOrder = getTiffByteOrder(byteOrder1);
-        setByteOrder(byteOrder);
-
-        final int tiffVersion = read2Bytes("tiffVersion", is, "Not a Valid TIFF File", getByteOrder());
-        if (tiffVersion != 42) {
-            throw new ImagingException("Unknown TIFF Version: " + tiffVersion);
-        }
+   private TiffHeader readTiffHeader(final InputStream is) throws ImagingException, IOException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java:476]: Formatting seems odd.

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hi @gwlucastrig 
Thank you for your updates. Please see my comments.

**Comment by gwlucastrig** [comment]
Added Javadoc and corrected formatting.  In order to support an additional JUnit test, I needed to add a new toBytes() method for longs.  This method wasn't required for reading BigTIFFs, but it was necessary to implement the JUnit test.  It will be useful in the future for any implementation that writes BigTIFFs (the current implementation only reads them).

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java (hunk 3 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java]: done

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java (hunk 6 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java]: done

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java (hunk 5 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:345]: done

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java (hunk 4 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:307]: done

File: src/main/java/org/apache/commons/imaging/common/BinaryFunctions.java (hunk 1 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/BinaryFunctions.java:192]: done

File: src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java [hunk 7]
```diff
@@ -16,27 +16,29 @@
...
 import java.io.InputStream;
 import java.nio.ByteOrder;
 import java.util.ArrayList;
 import java.util.List;
-
 import org.apache.commons.imaging.FormatCompliance;
 import org.apache.commons.imaging.ImagingException;
 import org.apache.commons.imaging.bytesource.ByteSource;
 import org.apache.commons.imaging.common.BinaryFileParser;
+import static org.apache.commons.imaging.common.BinaryFunctions.read2Bytes;
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java:28]: All statics should be grouped at the top. Not sure if Checkstyle can check for us though.

Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java:28]: I ran checkstyle before submitting and it didn't complain.  But I definitely agree that the statics should go first. I will modify this and submit a new commit tonight or tomorrow.

Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java]: done

**Review by garydgregory — APPROVED** [review]
LGTM

**Pull request merged**